src/Compiler/Driver/OptimizeInputs.fs

```diff
@@ -48,6 +51,288 @@ let GetInitialOptimizationEnv (tcImports: TcImports, tcGlobals: TcGlobals) =
+        {
+            mutable Phase1: Phase1Res option
+            mutable Phase2: Phase2Res option
+            mutable Phase3: Phase3Res option
+        }
+
+        member this.HasResult(phase: OptimizationPhase) =
+            match phase with
+            | OptimizationPhase.Phase1 -> this.Phase1 |> Option.isSome
+            | OptimizationPhase.Phase2 -> this.Phase2 |> Option.isSome
+            | OptimizationPhase.Phase3 -> this.Phase3 |> Option.isSome
+
+        static member Empty =
+            {
+                Phase1 = None
+                Phase2 = None
+                Phase3 = None
+            }
+
+    type Node =
+        {
+            Idx: int
+            Phase: OptimizationPhase
+        }
+
+        override this.ToString() = $"[{this.Idx}-{this.Phase}]"
+
+    let collectResults
+        (inputs: FileResultsComplete[])
+        : (CheckedImplFileAfterOptimization * ImplFileOptimizationInfo)[] * IncrementalOptimizationEnv =
+        let files =
+            inputs
+            |> Array.map
+                (fun {
+                         Phase1 = phase1
+                         Phase2 = _phase2
+                         Phase3 = phase3
+                     } ->
+                    let (_, _, implFileOptData, _), optimizeDuringCodeGen = phase1
+                    let _, implFile = phase3
+
+                    let implFile =
+                        {
+                            ImplFile = implFile
+                            OptimizeDuringCodeGen = optimizeDuringCodeGen
+                        }
+
+                    implFile, implFileOptData)
+
+        let lastFilePhase1Env =
+            inputs
+            |> Array.last
+            |> fun { Phase1 = phase1 } ->
+                let (optEnvPhase1, _, _, _), _ = phase1
+                optEnvPhase1
+
+        files, lastFilePhase1Env
+
+    let getPhase2Res (p: FileResults) = p.Phase2 |> Option.get
+
+    let getPhase3Res (p: FileResults) =
+        p.Phase3 |> Option.get |> (fun (env, _) -> env)
+
+    let optimizeFilesInParallel
+        (env0: IncrementalOptimizationEnv)
+        ((phase1, phase2, phase3): FilePhaseFuncs)
+        (files: CheckedImplFile list)
+        (ct: CancellationToken)
+        : (CheckedImplFileAfterOptimization * ImplFileOptimizationInfo)[] * IncrementalOptimizationEnv =
+        let files = files |> List.toArray
+
+        let firstNode =
+            {
+                Idx = 0
+                Phase = OptimizationPhase.Phase1
+            }
+
+        let results = files |> Array.map (fun _ -> FileResults.Empty)
+
+        let _lock = obj ()
+
+        let nodeCanBeProcessed { Idx = idx; Phase = phase } : bool =
+            let previousPhase =
+                match phase with
+                | OptimizationPhase.Phase1 -> None
+                | OptimizationPhase.Phase2 -> Some OptimizationPhase.Phase1
+                | OptimizationPhase.Phase3 -> Some OptimizationPhase.Phase2
+
+            lock _lock (fun () ->
```

Comment:
I do not see how the write side and read side of this _lock are connected to each other.
Either the writing part using the same _lock needs to increase it's scope, or maybe the _lock on the reading side (this code next to the comment) can be avoided alltogether.